engine/api/src/main/java/io/deephaven/engine/table/PartitionedTable.java

```diff
@@ -0,0 +1,199 @@
+ * tables.
+ * <p>
+ * Note that partitioned tables should {@link io.deephaven.engine.updategraph.NotificationQueue.Dependency depend} on
+ * and {@link io.deephaven.engine.liveness.LivenessManager#manage(LivenessReferent) manage} their
+ * {@link Table#isRefreshing() refreshing} constituents.
+ */
+public interface PartitionedTable extends LivenessNode, LogOutputAppendable {
+
+    /**
+     * Interface for proxies created by {@link #proxy()}.
+     */
+    interface Proxy extends TableOperations<Proxy, TableOperations> {
+
+        /**
+         * Get the PartitionedTable instance underlying this proxy.
+         *
+         * @return The underlying PartitionedTable instance
+         */
+        PartitionedTable target();
+
+        /**
+         * @return Whether this Proxy ensures matching partition keys for implicit joins of partitioned tables
+         * @see PartitionedTable#proxy(boolean, boolean)
+         */
+        boolean requiresMatchingKeys();
+
+        /**
+         * @return Whether this Proxy ensures non-overlapping join keys in constituent tables for proxied joins
+         * @see PartitionedTable#proxy(boolean, boolean)
+         */
+        boolean sanityChecksJoins();
+    }
+
+    /**
+     * Get the "raw" {@link Table partitioned table} underlying this PartitionedTable.
+     *
+     * @return The underlying {@link Table partitioned table}
+     */
+    Table table();
+
+    /**
+     * Get the names of all "key" columns that are part of {@code table().getDefinition()}. If there are no key columns,
+     * the result will be empty.
+     *
+     * @return The key column names
+     */
+    Set<String> keyColumnNames();
+
+    /**
+     * Get the name of the "constituent" column of {@link Table tables}.
+     *
+     * @return The constituent column name
+     */
+    String constituentColumnName();
+
+    /**
+     * Get a {@link TableDefinition definition} shared by or
+     * {@link TableDefinition#checkMutualCompatibility(TableDefinition) mutually compatible} with all constituent
+     * {@link Table tables}.
+     *
+     * @return The constituent definition
+     */
+    TableDefinition constituentDefinition();
+
+    /**
+     * <p>
+     * Can the constituents of the underlying {@link #table() partitioned table} change?
+     * <p>
+     * This is completely unrelated to whether the constituents themselves are {@link Table#isRefreshing() refreshing},
+     * or whether the underlying partitioned table is refreshing. Note that the underlying partitioned table
+     * <em>must</em> be refreshing if it contains any refreshing constituents.
+     * <p>
+     * PartitionedTables that specify {@code constituentChangesPermitted() == true} must be guaranteed to never change
+     * their constituents. Formally, it is expected that
+     * {@code table().getColumnSource(constituentColumnName()).isImmutable() == true}, that {@code table()} will never
```

Comment:
I'm going to take out the bit about immutability. I only care about *functional* immutability, which the rest of the sentence dictates.
While your statement is reasonable, I don't have any reason at to care care if columns other than the constituent change; this guarantee/assertion is really just for the semantics of `merge()`.
Please resolve if you're convinced. =)

Reply:
I was concerned about proxy usage of joins.  That if you modify the keys, the original table you generated no longer makes sense.  Presumably it would get removed from the result and be fine enough, but it does seem like you could get yourself into unpleasant situations.